pool.go

```diff
@@ -0,0 +1,241 @@
+package connection
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+)
+
+type ConnectionFactoryFunc func(addr string) (*Connection, error)
+
+type Pool struct {
+	Factory ConnectionFactoryFunc
+	Addrs   []string
+	Opts    PoolOptions
+
+	mu          sync.Mutex
+	connections []*Connection
+	connIndex   int
+
+	done chan struct{}
+
+	// WaitGroup to wait for all reconnect goroutines return
+	wg sync.WaitGroup
+
+	isClosed bool
+}
+
+// Pool - provides connections to the clients. It removes the connection from
+// the pool if it was closed and in the background tries to create and etablish
+// new connection so the pool will be full.
+func NewPool(factory ConnectionFactoryFunc, addrs []string, options ...PoolOption) (*Pool, error) {
```

Comment:
Do you know if there's a pool we could rely on in the OSS space somewhere? [HikariCP](https://github.com/brettwooldridge/HikariCP) is an example (in Java) which has been incredibly well optimized and tested, so I wish a Go variant existed.

Reply:
I'll check. 
I saw some, but the issue was that in "traditional" pool you get a connection, use it, and return it back to the pool. While here we don't return it. We also can get the same connection from the pool. Maybe the pool is not the proper name here...

Reply:
"pool" might not be the correct term here. We're looking for something which is part Factory and part gate.